### ISSUE TITLE
OpenTracing support

### ISSUE THREAD

**Issue by kr** [lib/pq#667]
Do maintainers think it would be reasonable to have this package create OpenTracing spans if a trace exists in the context?

See http://opentracing.io/ for background info.

If there's interest, I'd be happy to contribute code or whatever is necessary. (Otherwise, we will probably write a database/sql driver that wraps lib/pq to do this.)

I would try to do this in a way that avoids a hard dependency on the opentracing packages, so that users who are uninterested can simply ignore it, and don't have to encounter a new package dependency.

**Comment by kr** [comment]
This might even work in a way that is independent of OpenTracing per se. It might be a very small, generic tracing hook, like:

```go
func (d *Driver) SetTracer(trace func(context.Context, string) (finish func()))
```

And clients would use it something like this:

```go
driver := pq.Open(...)
driver.(*pq.Driver).SetTracer(func(ctx context.Context, opname string) func() {
    span, _ := StartSpanFromContext(ctx, opname)
    return span.Finish
})
sql.Register("tracingpq", driver)
```

**Comment by kr** [comment]
In some amazing future where the postgres backend *also* does OpenTracing, lib/pq might want to propagate tracing info to the backend. In that world, I would suggest just taking a hard dependency on OpenTracing and deprecating this generic hook.

**Comment by kr** [comment]
https://github.com/opentracing/specification/blob/master/semantic_conventions.md

There are also a few pieces of metadata that lib/pq would be in an excellent position to provide automatically, if committing to OpenTracing is reasonable:

- `db.instance`
- `db.statement`
- `db.type`
- `db.user`

**Comment by ainar-g** [comment]
This is the first time I hear about OpenTracing, and it sounds interesting, but as a package user I'd prefer my database driver not to have a dependency. If I ever use OpenTracing at work, I would probably want to create database-related spans myself. And, as you said, wrapping the driver is always an option.

**Comment by otto-md** [comment]
Have you seen https://github.com/ExpansiveWorlds/instrumentedsql ?

**Comment by kr** [comment]
> Have you seen https://github.com/ExpansiveWorlds/instrumentedsql ?

Yeah, that is a bit too kitchen-sinky for my taste, but it is an option.

> And, as you said, wrapping the driver is always an option.

I just got through writing a first pass at a wrapper, and it was not too bad.

However, there's one tag I've encountered that doesn't seem feasible to do in a wrapper, and that is the actual peer IP address. If a dns name resolves to multiple A or AAAA records, it can be useful to know which backend server the query actually ran on, for example to notice that a particular pg server is slower than the others. Currently, we record the hostname+port (as `peer.address`) but not the resolved IPv4 addr (which would be `peer.ipv4`).

**Comment by kr** [comment]
> the actual peer IP address

This could alternatively be done with a tiny helper function in lib/pq such as `func PeerAddress(driver.Conn) net.IP` or something along those lines.

**Comment by JelteF** [comment]
@kr I'm very interested in this as well. Is your wrapper public somewhere?

**Comment by kr** [comment]
It's not yet, but there's a good chance we'll make it public soon-ish. I don't have a specific timeframe though, I'm afraid.

**Comment by JelteF** [comment]
Sounds awesome! I'm familiar with how open sourcing internal stuff always
takes more time than you hope, mostly because of the need to clean it up.
Just to set my expectations somewhat correctly, what do you mean with
soonish? Couple of days, weeks or months?

**Comment by kr** [comment]
> what do you mean with soonish? Couple of days, weeks or months?

I don't know, sorry. 😬

**Comment by JelteF** [comment]
No worries. But we might build something ourselves then ;)

**Comment by kr** [comment]
> But we might build something ourselves then

Yeah I would encourage that. Our wrapper is 165 lines of boring (not subtle) glue code (including whitespace, comments, and documentation), so it's probably not even worth taking a new dependency for it.

**Comment by Monnoroch** [comment]
Hi all,

We're using this drived for CocroachDB which actually does support OpenTracing. How hard would it be to implement passing the OT headers to the server? I don't think the driver is the place to create spans and this should be left to the developer (or indeed to the driver wrapper), but passing the headers can only be implemented in a driver.

**Comment by GeertJohan** [comment]
For creating spans on calls to the DB, the driver itself doesn't need support. I agree with @ainar-g that pq shouldn't have it as dependency. In the OpenCensus ecosystem there is an SQL driver wrapper, [ocsql](https://github.com/basvanbeek/ocsql), which adds tracing support to existing drivers. It does however require support for the *Context methods. Perhaps OpenTracing could do the same? Issue #786 may be interesting to the readers of this issue.

**Comment by basvanbeek** [comment]
There is no need for a database driver to add tracing dependencies as a driver wrap like `ocsql` will be sufficient. No need to clutter codebase especially since there is more than just OpenTracing (ex. OpenCensus, Zipkin go, etc.)

Also I see there is confusion about the *Context methods as provided by the database/sql package. These methods will be callable and downgrade to the non context ones internally if a driver does not support context.Context. By having a proper tracing wrapper you don't need a driver to implement context.Context for tracing as long as the wrapper like `ocsql` is being called with the *Context functions so it can extract the current span from the active trace.

Only reason to ask for context.Context support is for deadline / cancellation purposes OR if the actual database contains instrumentation so the tracing SpanContext (trace identifiers) can be transported as metadata to queries.

**Comment by renannprado** [comment]
@kr I didn't try, but maybe this can help: https://github.com/gchaincl/sqlhooks

**Comment by batazor** [comment]
Is there any news on this?

**Comment by A-Kamaee** [comment]
I have the same question. Any update?

**Comment by arp242** [comment]
I'm gonna say that this is probably best handled at a higher level, like the aforementioned ocsql. That will work with any database driver.

That is, unless there is a compelling reason it needs to be included here (e.g. for things that can't be accessed at a higher level), but reading the above discussion no one mentioned that? I don't strictly mind adding something for it, but not clear if it's needed (or what's needed).
